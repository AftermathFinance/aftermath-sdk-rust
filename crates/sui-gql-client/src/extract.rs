//! Defines [`extract!`](crate::extract!) and its [`Error`].

/// Error for [`extract!`](crate::extract!).
#[derive(thiserror::Error, Debug)]
#[error("Missing data from response: {0}")]
pub struct Error(pub(crate) String);

impl Error {
    pub fn new(s: impl Into<String>) -> Self {
        Self(s.into())
    }
}

/// Helper for extracting data from GraphQL responses.
///
/// Designed specially to deal with the multitude of nested [`Option`]s commonly found in GQL
/// responses. The macro will generate an error string with the full path to the missing attribute.
///
/// # Example
///
/// ```no_run
/// # use af_sui_types::Address;
/// # use sui_gql_schema::{schema, scalars};
/// #
/// # #[derive(cynic::QueryVariables, Debug)]
/// # struct QueryVariables<'a> {
/// #     ch: Address,
/// #     vault: DynamicFieldName<'a>,
/// # }
/// #
/// # #[derive(cynic::InputObject, Debug)]
/// # struct DynamicFieldName<'a> {
/// #     #[cynic(rename = "type")]
/// #     type_: &'a str,
/// #     bcs: scalars::Base64<Vec<u8>>,
/// # }
/// #
/// #[derive(cynic::QueryFragment, Debug)]
/// #[cynic(variables = "QueryVariables")]
/// struct Query {
///     #[arguments(address: $ch)]
///     object: Option<Object>,
/// }
///
/// #[derive(cynic::QueryFragment, Debug)]
/// #[cynic(variables = "QueryVariables")]
/// struct Object {
///     #[arguments(name: $vault)]
///     dynamic_field: Option<DynamicField>,
/// }
///
/// #[derive(cynic::QueryFragment, Debug)]
/// struct DynamicField {
///     value: Option<DynamicFieldValue>,
/// }
///
/// #[derive(cynic::InlineFragments, Debug)]
/// enum DynamicFieldValue {
///     MoveValue(MoveValue),
///     #[cynic(fallback)]
///     Unknown
/// }
///
/// #[derive(cynic::QueryFragment, Debug)]
/// struct MoveValue {
///     #[cynic(rename = "type")]
///     type_: MoveType,
///     bcs: scalars::Base64<Vec<u8>>,
///     __typename: String,
/// }
///
/// #[derive(cynic::QueryFragment, Debug)]
/// struct MoveType {
///     repr: String,
/// }
///
/// use sui_gql_client::extract;
///
/// // Could be obtained in practice from `sui_gql_client::GraphQlResponseExt::try_into_data`
/// let data: Option<Query> = None;
/// let df_value: MoveValue = extract!(
///     data?.object?.dynamic_field?.value?.as_variant(DynamicFieldValue::MoveValue)
/// );
/// # color_eyre::eyre::Ok(())
/// ```
#[macro_export]
macro_rules! extract {
    ($ident:ident $($tt:tt)*) => {{
        $crate::extract!( @attributes [$ident][$($tt)*] -> {
            $ident
        } )
    }};

    (@attributes [$($path:tt)*][] -> { $($expr:tt)* } ) => { $($expr)* };

    (@attributes [$($path:tt)*][.as_variant($($var:tt)+) $($tt:tt)*] -> { $($expr:tt)* } ) => {
        $crate::extract!(@attributes
            [$($path)*.as_variant($($var)*)][$($tt)*] -> {{
                let value = $($expr)*;
                let $($var)+(var) = value else {
                    let msg = stringify!($($path)*.as_variant($($var)*));
                    return Err($crate::extract::Error::new(msg).into());
                };
                var
            }}
        )
    };

    (@attributes [$($path:tt)*][? $($tt:tt)*] -> { $($expr:tt)* } ) => {
        $crate::extract!(@attributes
            [$($path)*][$($tt)*] -> {
                $($expr)*
                .ok_or_else(|| {
                    let msg = stringify!($($path)*);
                    $crate::extract::Error::new(msg)
                })?
            }
        )
    };

    (@attributes [$($path:tt)*][.$ident:ident $($tt:tt)*] -> { $($expr:tt)* } ) => {
        $crate::extract!(@attributes
            [$($path)*.$ident][$($tt)*] -> {
                $($expr)*
                .$ident
            }
        )
    };
}

#[cfg(test)]
mod tests {
    use af_sui_types::Address;
    use sui_gql_schema::{scalars, schema};

    #[derive(cynic::QueryVariables, Debug)]
    #[expect(dead_code, reason = "for cynic")]
    struct Variables<'a> {
        ch: Address,
        vault: DynamicFieldName<'a>,
    }

    #[derive(cynic::InputObject, Debug)]
    struct DynamicFieldName<'a> {
        #[cynic(rename = "type")]
        type_: &'a str,
        bcs: scalars::Base64<Vec<u8>>,
    }

    #[derive(cynic::QueryFragment, Debug)]
    #[cynic(variables = "Variables")]
    struct Query {
        #[arguments(address: $ch)]
        object: Option<Object>,
    }

    #[derive(cynic::QueryFragment, Debug)]
    #[cynic(variables = "Variables")]
    struct Object {
        #[arguments(name: $vault)]
        dynamic_field: Option<DynamicField>,
    }

    #[derive(cynic::QueryFragment, Debug)]
    struct DynamicField {
        value: Option<DynamicFieldValue>,
    }

    #[derive(cynic::InlineFragments, Debug)]
    enum DynamicFieldValue {
        MoveValue(MoveValue),
        #[cynic(fallback)]
        Unknown,
    }

    #[derive(cynic::QueryFragment, Debug)]
    struct MoveValue {
        __typename: String,
    }

    #[test]
    fn error_display() {
        fn extract_data(data: Option<Query>) -> Result<MoveValue, super::Error> {
            Ok(extract!(data?
                .object?
                .dynamic_field?
                .value?
                .as_variant(DynamicFieldValue::MoveValue)))
        }

        let data = Some(Query {
            object: Some(Object {
                dynamic_field: Some(DynamicField {
                    value: Some(DynamicFieldValue::Unknown),
                }),
            }),
        });

        let res = extract_data(data);
        assert!(res.is_err());
        insta::assert_snapshot!(res.err().unwrap(), @"Missing data from response: data.object.dynamic_field.value.as_variant(DynamicFieldValue::MoveValue)");

        let res = extract_data(None);
        assert!(res.is_err());
        insta::assert_snapshot!(res.err().unwrap(), @"Missing data from response: data");

        let res = extract_data(Some(Query {
            object: Some(Object {
                dynamic_field: None,
            }),
        }));
        assert!(res.is_err());
        insta::assert_snapshot!(res.err().unwrap(), @"Missing data from response: data.object.dynamic_field");
    }
}
