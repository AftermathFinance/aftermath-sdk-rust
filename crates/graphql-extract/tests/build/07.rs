struct Query {
    object: Option<Object>,
}

struct Object {
    version: u64,
    dynamic_field: Option<DynamicField>,
}

struct DynamicField {
    value: Option<DynamicFieldValue>,
}

enum DynamicFieldValue {
    MoveValue(MoveValue),
    Unknown,
}

struct MoveValue {
    type_: MoveType,
    bcs: String,
}

struct MoveType {
    repr: String,
}

type Item = Result<(MoveType, String), &'static str>;

fn extract(data: Option<Query>) -> Result<(u64, impl Iterator<Item = Item>), &'static str> {
    use graphql_extract::extract;
    use DynamicFieldValue::MoveValue;

    extract!(data => {
        object? {
            version
            dynamic_field {
                value? {
                    ... on MoveValue {
                        type_
                        bcs
                    }
                    node
                }
            }
        }
    });
    Ok((version, nodes))
}

fn main() {}
