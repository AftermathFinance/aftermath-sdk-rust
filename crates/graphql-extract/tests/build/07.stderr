error: Only a single '... on Variant' match is supported within the same braces
  --> tests/build/07.rs:38:24
   |
38 |                   value? {
   |  ________________________^
39 | |                     ... on MoveValue {
40 | |                         type_
41 | |                         bcs
42 | |                     }
43 | |                     node
44 | |                 }
   | |_________________^

error[E0425]: cannot find value `version` in this scope
  --> tests/build/07.rs:48:9
   |
48 |     Ok((version, nodes))
   |         ^^^^^^^ not found in this scope

error[E0425]: cannot find value `nodes` in this scope
  --> tests/build/07.rs:48:18
   |
48 |     Ok((version, nodes))
   |                  ^^^^^ not found in this scope

warning: unused import: `DynamicFieldValue::MoveValue`
  --> tests/build/07.rs:32:9
   |
32 |     use DynamicFieldValue::MoveValue;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default
